### PR TITLE
Add javadoc and tests for functional Topic class

### PR DESCRIPTION
**Pull request by mziccard**
This PR adds unit tests and javadoc to functional `Topic` class.

### CODE REVIEW

File: gcloud-java-pubsub/src/test/java/com/google/cloud/pubsub/TopicTest.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+public class TopicTest {
+
+  private static final String NAME = "topic";
+  private static final TopicInfo TOPIC_INFO = TopicInfo.of(NAME);
+
+  private PubSub serviceMockReturnsOptions = createStrictMock(PubSub.class);
+  private PubSubOptions mockOptions = createMock(PubSubOptions.class);
```
Comment (mziccard) [inline gcloud-java-pubsub/src/test/java/com/google/cloud/pubsub/TopicTest.java]: Codacy Issue found: [Private field 'mockOptions' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2613573683/issues/source?bid=3123109&fileBranchId=3311984#l49)

File: gcloud-java-pubsub/src/test/java/com/google/cloud/pubsub/TopicTest.java [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+public class TopicTest {
+
+  private static final String NAME = "topic";
+  private static final TopicInfo TOPIC_INFO = TopicInfo.of(NAME);
+
+  private PubSub serviceMockReturnsOptions = createStrictMock(PubSub.class);
```
Comment (mziccard) [inline gcloud-java-pubsub/src/test/java/com/google/cloud/pubsub/TopicTest.java]: Codacy Issue found: [Private field 'serviceMockReturnsOptions' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2613573683/issues/source?bid=3123109&fileBranchId=3311984#l48)

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/Topic.java [hunk 3]
```diff
@@ -73,70 +82,173 @@ public Builder toBuilder() {
...
 
   @Override
-  public boolean equals(Object obj) {
-    if (this == obj) {
+  public final boolean equals(Object obj) {
+    if (obj == this) {
       return true;
     }
-    if (obj == null || getClass() != obj.getClass()) {
+    if (obj == null || !obj.getClass().equals(Topic.class)) {
```
Comment (jgeewax) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/Topic.java]: Why the switch here?

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/Topic.java [hunk 4]
```diff
@@ -73,70 +82,173 @@ public Builder toBuilder() {
...
       return false;
     }
     Topic other = (Topic) obj;
-    return Objects.equals(name(), other.name()) && Objects.equals(options, other.options);
+    return baseEquals(other) && Objects.equals(options, other.options);
```
Comment (jgeewax) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/Topic.java]: Where is `baseEquals` defined?

Comment (jgeewax) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/Topic.java]: nevermind -- found it at the bottom :)

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/TopicInfo.java [hunk 5]
```diff
@@ -98,15 +98,16 @@ public int hashCode() {
...
   @Override
   public boolean equals(Object obj) {
-    if (this == obj) {
-      return true;
-    }
-    if (obj == null || !obj.getClass().equals(this.getClass())) {
-      return false;
-    }
-    return Objects.equals(name, ((TopicInfo) obj).name);
+    return obj == this
```
Comment (jgeewax) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/TopicInfo.java]: i definitely prefer the conditionals here. this boolean logic can get tricky...

Comment (mziccard) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/TopicInfo.java]: The reason I changed is just for consistency with other classes (see [BlobInfo](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-storage/src/main/java/com/google/cloud/storage/BlobInfo.java#L640) or [ZoneInfo](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-dns/src/main/java/com/google/cloud/dns/ZoneInfo.java#L295)). If you have a preference for conditionals I don't mind changing though we should change it everywhere to be consistent.

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/Topic.java (hunk 3 above)
Comment (mziccard) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/Topic.java]: I changed only for consistency with other classes across `gcloud-java` (see [Zone](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-dns/src/main/java/com/google/cloud/dns/Zone.java#L203) or [Blob](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-storage/src/main/java/com/google/cloud/storage/Blob.java#L532) for example).

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/TopicInfo.java [hunk 6]
```diff
@@ -98,15 +98,16 @@ public int hashCode() {
...
+        || obj != null
```
Comment (lesv) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/TopicInfo.java]: @jgeewax I understand your concerns, my inclination was to ask for additional parens, but I looked closer and at two referenced methods and am inclined to say Ok w/ this (and the prior one), unless you object.

**Comment by lesv** [comment]
LGTM - @jgeewax as I mentioned earlier, I'm inclined to approve this, unless you object.
@mziccard please give jjg until 3pm PT to object, then Ok to merge.

**Pull request merged**